crates/llms/src/responses.rs

```
@@ -0,0 +1,93 @@
+use tracing_futures::Instrument;
+
+use crate::chat::nsql::SqlGeneration;
+
+#[derive(Debug, Snafu)]
+#[snafu(visibility(pub))]
+pub enum Error {
+    #[snafu(display(
+        "Failed to load the model.\nAn error occurred: {source}\nReport a bug on GitHub: https://github.com/spiceai/spiceai/issues"
+    ))]
+    FailedToLoadModel {
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+    #[snafu(display(
+        "Failed to run the model.\nAn error occurred: {source}\nReport a bug on GitHub: https://github.com/spiceai/spiceai/issues"
+    ))]
+    FailedToRunModel {
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+    #[snafu(display(
+        "Failed to check the status of the model. An error occurred: {source}. Verify the model configuration and try again."
+    ))]
+    HealthCheckError {
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[async_trait]
+pub trait Responses: Sync + Send {
+    fn as_sql(&self) -> Option<&dyn SqlGeneration> {
+        None
+    }
+    async fn run(&self, prompt: String) -> Result<Option<String>> {
+        let span = tracing::Span::current();
+
+        async move {
+            let req = CreateResponseArgs::default()
+                .input(prompt)
+                .build()
+                .boxed()
+                .context(FailedToLoadModelSnafu)?;
+
+            let resp = self
+                .responses_request(req)
+                .await
+                .boxed()
+                .context(FailedToRunModelSnafu)?;
+
+            Ok(resp.output_text)
+        }
+        .instrument(span)
+        .await
+    }
+
+    async fn health(&self) -> Result<()>;
+    async fn stream<'a>(
+        &self,
+        prompt: String,
+    ) -> Result<Pin<Box<dyn Stream<Item = Result<Option<String>>> + Send>>> {
+        let resp = self.run(prompt).await;
+        Ok(Box::pin(stream! { yield resp }))
+    }
```

```suggestion
   fn stream(self: Arc<Self>, prompt: impl Into<String>)
    -> Result<Pin<Box<dyn Stream<Item = Result<Option<String>>> + Send>>> {
    let prompt = prompt.into();
    Box::pin(stream::once(async move { self.run(prompt).await }))
}
```

1. The function that returns the stream doesn't need to be async if you change `self` to be `Arc<Self>` then you can move it into the async closure without lifetime issues.
2. `stream::once` is more idiomatic than using `stream!`
3. The stream function itself doesn't need to return an error, the error can be returned in the stream itself (i.e. Item = Result<..>)

---

This runs into the lifetime issue of `Self` not living long enough